### PR TITLE
Dasharo (UEFI) v0.1.0 for QEMU Q35

### PR DESCRIPTION
**Pull request by maheshtammisetti**
In order to run the Dasharo OVMF in QEMU, please follow the given steps in order to build the Dasharo firmware image from the EDK2 repository and then run the firmware image using the QEMU emulator.

1. Clone the official Dasharo EDK2 repository to your local instance from https://github.com/Dasharo/edk2.git with git or downloading the source code from github. 
2. Once it is done, please follow the instructions from the `README.md` in order to prepare your environment for the building of the OVMF image. 
3. In order to build the UEFI firmware image it can be done in two ways, one way is to give the flags with `-D SMM_REQUIRE ` (SMM was actually enabled while building the firmware image, but the flags differ with feature to feature, so please check the OvmfPkgX64.dsc file). And second way is to set the flags in OvmfPkgX64.dsc and then build the firmware image. For the purpose of this documentation, I have preferred to set the features in "OvmfPkgX64.dsc" directly.
4. Check the following build command ` build -D DEBUG_ON_SERIAL_PORT -D BOOTLOADER=COREBOOT -a IA32 -a X64 -t GCC5 -b DEBUG -p OvmfPkg/OvmfPkgX64.dsc `. 

-  `DEBUG_ON_SERIAL_PORT`  = The following flag must be given for the "DEBUG" build as the output is redirected to the I/O port `0x402` to qemu monitor. We will be able to see the debug messages on the given debug I/O port. (0x402 will be later utilized in the qemu invocation further in this documentation). 
5. Once the build is completed, the OVMF firmware image can be found at ` edk2/Build/Ovmf/DEBUG_GCC5/Ovmf.fd `. 
6. Please verify your qemu installation, and invoke the qemu emulator with the built OVMF firmware image as given below.

- ` qemu-system-x86_64 -drive if=pflash,format=raw,file=Build/OvmfX64/DEBUG_GCC5/FV/OVMF.fd -nographic -debugcon file:debug.log -global isa-debugcon.iobase=0x402 -global ICH9-LPC.disable_s3=1 -net none -machine q35,smm=on `
- " The ` -drive` parameter indicates that the device is a pflash with the firmware image file pointing out to the built `OVMF.fd` image. 
- `-nographic` indicates to start the qemu emulator without any graphical output. This is useful while testing in the environment like docker. 
-  `-debugcon file:debug.log -global isa-debugcon.iobase=0x402` default OVMF build writes debug messages to IO port 0x402.  The following qemu command line options save them in the file called debug.log.
-  `-global ICH9-LPC.disable_s3=1` SMM is put to use in the S3 suspend and resume infrastructure, and in the UEFI variable driver stack. Similarly, a pflash-backed variable store is a requirement. As SMM was enabled while building the firmware image.
-  `-net none` disables the OptionROM execution which will interfere with the firmware image, if not disabled.
-  `-machine q35,smm=on` For SMM to work, only Q35 machines are supported hence the machine type.
7. After executing the above qemu command, one should be able to see the UEFI built-in shell and should be able to get to the BIOS selection area.
8. The features which are enabled in the `OvmfPkgX64.dsc` can be verified in the `BIOS menu` and also at the `Device Manager` section and the Dasharo features can be verified at the `Dasharo System Features` section.

### CODE REVIEW

**Comment by pietrushnic** [comment]
@maheshtammisetti PR description is not correct way for contributing documentation. Please propose changes [here](https://github.com/Dasharo/docs)

**Comment by pietrushnic** [comment]
I guess official name should be something like Dasharo (UEFI) v0.1.0 for QEMU Q35 or something like that. So in Supporter Hardware we should add QEMU Q35 and try to fill as many section as we can described [here](https://github.com/Dasharo/docs#supported-hardware)

File: OvmfPkg/OvmfPkgX64.dsc [hunk 1]
```diff
@@ -450,7 +482,7 @@
   gEfiMdeModulePkgTokenSpaceGuid.PcdConOutUgaSupport|FALSE
   gEfiMdeModulePkgTokenSpaceGuid.PcdInstallAcpiSdtProtocol|TRUE
 !ifdef $(CSM_ENABLE)
-  gUefiOvmfPkgTokenSpaceGuid.PcdCsmEnable|TRUE
+  gUefiOvmfPkgTokenSpaceGuid.PcdCsmEnable|FALSE
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: Please revert this change. We do not need CSM

File: OvmfPkg/OvmfPkgX64.dsc [hunk 2]
```diff
@@ -608,6 +640,10 @@
 
 !if $(TPM_ENABLE) == TRUE
   gEfiSecurityPkgTokenSpaceGuid.PcdTpmInstanceGuid|{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
+  ### TPM Initialization.
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpm2InitializationPolicy|1
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpm2SelfTestPolicy|1
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpmInitializationPolicy|1
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: These PCDs should not be touched. Please revert

File: OvmfPkg/OvmfPkgX64.dsc [hunk 3]
```diff
@@ -848,7 +926,7 @@
 !ifdef $(CSM_ENABLE)
   OvmfPkg/Csm/BiosThunk/VideoDxe/VideoDxe.inf {
     <LibraryClasses>
-      PcdLib|MdePkg/Library/DxePcdLib/DxePcdLib.inf
+     PcdLib|MdePkg/Library/DxePcdLib/DxePcdLib.inf
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: Useless change, please revert.

File: OvmfPkg/OvmfPkgX64.fdf [hunk 4]
```diff
@@ -119,6 +119,8 @@ INF  OvmfPkg/Sec/SecMain.inf
 
 INF  RuleOverride=RESET_VECTOR OvmfPkg/ResetVector/ResetVector.inf
 
+
+
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.fdf:123]: Please do not add empty lines like that

File: OvmfPkg/OvmfPkgX64.dsc [hunk 5]
```diff
@@ -28,11 +28,20 @@
...
+  DEFINE TPM_ENABLE              = TRUE
+  DEFINE TPM_CONFIG_ENABLE       = TRUE
+  DEFINE SATA_PASSWORD_ENABLE    = TRUE
+  DEFINE OPAL_PASSWORD_ENABLE    = TRUE
+  DEFINE LOAD_OPTION_ROMS        = FLASE
+  DEFINE DASHARO_SYSTEM_FEATURES_ENABLE = TRUE
+  DEFINE USE_CBMEM_FOR_CONSOLE   = FALSE
+  DEFINE ABOVE_4G_MEMORY         = FALSE
+  DEFINE DISABLE_MTRR_PROGRAMMING = TRUE
+  DEFINE IOMMU_ENABLE            = FALSE
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc:42]: These will not be used by QEMU/OVMF, please remove

File: OvmfPkg/OvmfPkgX64.dsc [hunk 6]
```diff
@@ -28,11 +28,20 @@
...
+  DEFINE SECURE_BOOT_ENABLE      = TRUE
+  DEFINE SMM_REQUIRE             = TRUE
   DEFINE SOURCE_DEBUG_ENABLE     = FALSE
-  DEFINE TPM_ENABLE              = FALSE
-  DEFINE TPM_CONFIG_ENABLE       = FALSE
+  DEFINE TPM_ENABLE              = TRUE
+  DEFINE TPM_CONFIG_ENABLE       = TRUE
+  DEFINE SATA_PASSWORD_ENABLE    = TRUE
+  DEFINE OPAL_PASSWORD_ENABLE    = TRUE
+  DEFINE LOAD_OPTION_ROMS        = FLASE
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: This will not be used by QEMU/OVMF, please remove

**Review by miczyg1 — CHANGES_REQUESTED** [review]

File: OvmfPkg/OvmfPkgX64.dsc (hunk 6 above)
Comment (maheshtammisetti) [reply OvmfPkg/OvmfPkgX64.dsc]: Should I remove the LOAD_OPTIONS_ROMS line or all the lines?

File: OvmfPkg/OvmfPkgX64.dsc (hunk 5 above)
Comment (maheshtammisetti) [reply OvmfPkg/OvmfPkgX64.dsc:42]: I will remove those four lines.

File: OvmfPkg/OvmfPkgX64.dsc (hunk 6 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]: I have marked only the LOAD_OPTION_ROMS, but GItHub likes to preview a little bit more. In short:

```suggestion
```

**Comment by pietrushnic** [comment]
@maheshtammisetti in release notes of v0.1.0 we would have to clearly explain which Dasharo features work and which not. We should also consider what are difference with coreboot upstream config for Q35, if there is any for coreboot with UEFI payload.

@macpijan I guess we would like to assign to @maheshtammisetti task to get through release process for this binary. My vision is that people can sense some look and feel of Dasharo without running it on real hardware. This will be entry point for those who would like to learn Dasharo.

File: OvmfPkg/OvmfPkgX64.dsc (hunk 1 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]: @maheshtammisetti ?

File: OvmfPkg/OvmfPkgX64.dsc (hunk 3 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]: @maheshtammisetti ?

File: OvmfPkg/OvmfPkgX64.dsc (hunk 2 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]: I meant remove the lines that you added @maheshtammisetti

File: OvmfPkg/OvmfPkgX64.dsc (hunk 3 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]: @maheshtammisetti ?

Comment (maheshtammisetti) [reply OvmfPkg/OvmfPkgX64.dsc]: Will be removing that entire line.

File: OvmfPkg/OvmfPkgX64.dsc [hunk 7]
```diff
@@ -707,6 +747,21 @@
 !endif
   }
 
+### SETUP_PASSWORD
+
+!if $(SECURE_BOOT_ENABLE) == TRUE
+  SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigDxe.inf
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: The `SecureBootConfigDxe` is already present in the file, please remove all duplicates that you added

File: OvmfPkg/OvmfPkgX64.dsc [hunk 8]
```diff
@@ -729,22 +784,21 @@
...
       NULL|MdeModulePkg/Library/BootMaintenanceManagerUiLib/BootMaintenanceManagerUiLib.inf
-!ifdef $(CSM_ENABLE)
-      NULL|OvmfPkg/Csm/LegacyBootManagerLib/LegacyBootManagerLib.inf
-      NULL|OvmfPkg/Csm/LegacyBootMaintUiLib/LegacyBootMaintUiLib.inf
-!endif
+    <PcdsFixedAtBuild>
+      gDasharoSystemFeaturesTokenSpaceGuid.PcdShowMenu|$(DASHARO_SYSTEM_FEATURES_ENABLE)
   }
+  
+  MdeModulePkg/Application/BootManagerMenuApp/BootManagerMenuApp.inf
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc:797]: BootManagerMenuApp is not present in FDF file

File: OvmfPkg/OvmfPkgX64.dsc [hunk 9]
```diff
@@ -976,4 +1049,20 @@
...
       Tpm12DeviceLib|SecurityPkg/Library/Tpm12DeviceLibDTpm/Tpm12DeviceLibDTpm.inf
   }
+  SecurityPkg/Tcg/TcgConfigDxe/TcgConfigDxe.inf {
+    <LibraryClasses>
+      Tpm12DeviceLib|SecurityPkg/Library/Tpm12DeviceLibDTpm/Tpm12DeviceLibDTpm.inf
+  }
+
+
+!if $(OPAL_PASSWORD_ENABLE) == TRUE
+  SecurityPkg/Tcg/Opal/OpalPassword/OpalPasswordDxe.inf {
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc:1061]: `OpalPasswordDxe` is not present in FDF

File: OvmfPkg/OvmfPkgX64.dsc (hunk 7 above)
Comment (maheshtammisetti) [reply OvmfPkg/OvmfPkgX64.dsc]: Sure, working on it right away.

File: OvmfPkg/OvmfPkgX64.dsc [hunk 10]
```diff
@@ -707,6 +747,21 @@
 !endif
   }
 
+### SETUP_PASSWORD
+
+!if $(SECURE_BOOT_ENABLE) == TRUE
+  SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigDxe.inf
+  SecurityPkg/EnrollFromDefaultKeysApp/EnrollFromDefaultKeysApp.inf
+  SecurityPkg/VariableAuthenticated/SecureBootDefaultKeysDxe/SecureBootDefaultKeysDxe.inf
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: EnrollFromDefaultKeysApp and SecureBootDefaultKeysDxe are missing in FDF

File: OvmfPkg/OvmfPkgX64.dsc [hunk 11]
```diff
@@ -183,6 +198,16 @@
...
   IntrinsicLib|CryptoPkg/Library/IntrinsicLib/IntrinsicLib.inf
+#
+#OPAL_PASSWORD
+#
+
+!if $(OPAL_PASSWORD_ENABLE) == TRUE
+  TcgStorageCoreLib|SecurityPkg/Library/TcgStorageCoreLib/TcgStorageCoreLib.inf
+  TcgStorageOpalLib|SecurityPkg/Library/TcgStorageOpalLib/TcgStorageOpalLib.inf
+!endif
+  S3BootScriptLib|MdePkg/Library/BaseS3BootScriptLibNull/BaseS3BootScriptLibNull.inf
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: S3BootScriptLib is already present int the DSC file, please remove

File: OvmfPkg/OvmfPkgX64.dsc [hunk 12]
```diff
@@ -164,12 +168,23 @@
...
   LockBoxLib|OvmfPkg/Library/LockBoxLib/LockBoxBaseLib.inf
 !endif
   CustomizedDisplayLib|MdeModulePkg/Library/CustomizedDisplayLib/CustomizedDisplayLib.inf
   FrameBufferBltLib|MdeModulePkg/Library/FrameBufferBltLib/FrameBufferBltLib.inf
 
+#
+#Misc
+#
+
+  LockBoxLib|MdeModulePkg/Library/LockBoxNullLib/LockBoxNullLib.inf
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]:
```suggestion
!if $(SMM_REQUIRE) == FALSE
  LockBoxLib|OvmfPkg/Library/LockBoxLib/LockBoxBaseLib.inf
!else
  LockBoxLib|MdeModulePkg/Library/LockBoxNullLib/LockBoxNullLib.inf
!endif
  CustomizedDisplayLib|MdeModulePkg/Library/CustomizedDisplayLib/CustomizedDisplayLib.inf
  FrameBufferBltLib|MdeModulePkg/Library/FrameBufferBltLib/FrameBufferBltLib.inf
```

File: OvmfPkg/OvmfPkgX64.dsc [hunk 13]
```diff
@@ -297,7 +322,9 @@
   BaseCryptLib|CryptoPkg/Library/BaseCryptLib/PeiCryptLib.inf
   Tpm12DeviceLib|SecurityPkg/Library/Tpm12DeviceLibDTpm/Tpm12DeviceLibDTpm.inf
   Tpm2DeviceLib|SecurityPkg/Library/Tpm2DeviceLibDTpm/Tpm2DeviceLibDTpm.inf
+  Tcg2PhysicalPresenceLib|SecurityPkg/Library/PeiTcg2PhysicalPresenceLib/PeiTcg2PhysicalPresenceLib.inf
 !endif
+  PerformanceLib|MdeModulePkg/Library/PeiPerformanceLib/PeiPerformanceLib.inf
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: PerformanceLib is already present in this file. If we don't want to use to use FPDT we should remove this line. If we want to use FPDT then we have to remove the NULL PerformanceLib and add:

```
  MdeModulePkg/Universal/Acpi/FirmwarePerformanceDataTablePei/FirmwarePerformancePei.inf
  MdeModulePkg/Universal/Acpi/FirmwarePerformanceDataTableDxe/FirmwarePerformanceDxe.inf
```

To appropriate sections in DSC and FDF

File: OvmfPkg/OvmfPkgX64.dsc [hunk 14]
```diff
@@ -835,6 +889,25 @@
...
+    <LibraryClasses>
+    BaseCryptLib|CryptoPkg/Library/BaseCryptLib/BaseCryptLib.inf
+  }
+
+  #
+  # PKCS7 Verification
+  #
+  SecurityPkg/Pkcs7Verify/Pkcs7VerifyDxe/Pkcs7VerifyDxe.inf
+
+
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc:915]: I don;t see SmbiosMeasurementDxe anywhere in DSC and FDF but is present in UEFIPayloadPkg

File: OvmfPkg/OvmfPkgX64.dsc [hunk 15]
```diff
@@ -976,4 +1049,20 @@
...
+!if $(OPAL_PASSWORD_ENABLE) == TRUE
+  SecurityPkg/Tcg/Opal/OpalPassword/OpalPasswordDxe.inf {
+    <LibraryClasses>
+    Tpm2DeviceLib|SecurityPkg/Library/Tpm2DeviceLibRouter/Tpm2DeviceLibRouterDxe.inf
+  }
+!endif
+!endif
+
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc:1067]: Missing other storages in DSC and FDF like:

```
  MdeModulePkg/Bus/Pci/SdMmcPciHcDxe/SdMmcPciHcDxe.inf
  MdeModulePkg/Bus/Sd/EmmcDxe/EmmcDxe.inf
  MdeModulePkg/Bus/Sd/SdDxe/SdDxe.inf
```

**Review by miczyg1 — CHANGES_REQUESTED** [review]

**Comment by miczyg1** [comment]
@maheshtammisetti please address the discussions so we can merge the PR

**Comment by miczyg1** [comment]
@maheshtammisetti why did you force push the branch? Now I have trouble to find changes that you introduced...

Were there merge conflicts?

File: OvmfPkg/OvmfPkgX64.dsc (hunk 3 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]: No... I didn't ask to remove the line! You were supposed to revert the change introduced in this line (i.e. the unnecessary indentation change).  Thankfully you did it.

File: OvmfPkg/OvmfPkgX64.dsc [hunk 16]
```diff
@@ -795,6 +848,12 @@
 !endif
   OvmfPkg/QemuRamfbDxe/QemuRamfbDxe.inf
   OvmfPkg/VirtioGpuDxe/VirtioGpu.inf
+  
+  #
+  # SMBIOS Support
+  #
+  MdeModulePkg/Universal/SmbiosDxe/SmbiosDxe.inf
+  MdeModulePkg/Universal/SmbiosMeasurementDxe/SmbiosMeasurementDxe.inf
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]: These modules are repeated twice, please remove...

**Comment by miczyg1** [comment]
@maheshtammisetti a few comments remain unaddressed, please fix them.

File: MdeModulePkg/Universal/SecurityStubDxe/Defer3rdPartyImageLoad.c [hunk 17]
```diff
@@ -25,7 +25,7 @@ typedef struct {
 } DEFERRED_3RD_PARTY_IMAGE_TABLE;
 
 BOOLEAN                          mImageLoadedAfterEndOfDxe   = FALSE;
-BOOLEAN                          mEndOfDxe                   = TRUE;
+BOOLEAN                          mEndOfDxe                   = FALSE;
```
Comment (miczyg1) [inline MdeModulePkg/Universal/SecurityStubDxe/Defer3rdPartyImageLoad.c:28]: We should not modify this code. There must be a different way/solution to fix the OptionROM issue. mEndOfDxe is probably being set to FALSE by an event. Search for this event and try to understand when it is set

**Review by miczyg1 — CHANGES_REQUESTED** [review]

File: MdeModulePkg/Universal/SecurityStubDxe/Defer3rdPartyImageLoad.c (hunk 17 above)
Comment (miczyg1) [reply MdeModulePkg/Universal/SecurityStubDxe/Defer3rdPartyImageLoad.c:28]: Ok actually I see it was modified before https://github.com/Dasharo/edk2/commit/976412e7dd778ce281c5a138cb00dbe54a293e67
So it is a correct solution after all.

File: OvmfPkg/OvmfPkgX64.dsc (hunk 1 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]:
```suggestion
  gUefiOvmfPkgTokenSpaceGuid.PcdCsmEnable|TRUE
```

File: OvmfPkg/OvmfPkgX64.dsc [hunk 18]
```diff
@@ -608,7 +652,12 @@
 
 !if $(TPM_ENABLE) == TRUE
   gEfiSecurityPkgTokenSpaceGuid.PcdTpmInstanceGuid|{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
+  ### TPM Initialization.
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpm2InitializationPolicy|1
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpm2SelfTestPolicy|1
+  gEfiSecurityPkgTokenSpaceGuid.PcdTpmInitializationPolicy|1
```
Comment (miczyg1) [inline OvmfPkg/OvmfPkgX64.dsc]:
```suggestion
```

File: OvmfPkg/OvmfPkgX64.dsc (hunk 16 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]:
```suggestion
```

File: OvmfPkg/OvmfPkgX64.dsc (hunk 2 above)
Comment (miczyg1) [reply OvmfPkg/OvmfPkgX64.dsc]: Removed

**Review by miczyg1 — APPROVED** [review]

**Comment by miczyg1** [comment]
I have cleaned up and fixed the reamining comments. Merging
